module/zfs/spa.c

```diff
@@ -5439,8 +5439,17 @@ spa_import(char *pool, nvlist_t *config, nvlist_t *props, uint64_t flags)
 	if (spa_writeable(spa)) {
 		/*
 		 * Update the config cache to include the newly-imported pool.
+		 * We need to give away namespace lock before config update.
+		 * Since an export on this pool may be issued in parallel, It's
+		 * important to take a reference on spa before dropping the lock
 		 */
+		spa_open_ref(spa, FTAG);
+		mutex_exit(&spa_namespace_lock);
```

Comment:
This opens up a race conditions where other concurrent operations can occur. For example, an export/destroy could happen at the same time which might leave the pool in a strange state or result in a panic.

Reply:
Isn't taking an reference to spa enough to prevent the pool from getting exported in parallel or am I missing something here?

Reply:
It will prevent the export but it doesn't mean that the state of the pool is as expected. For example, I noticed that we call `zvol_remove_minors`  after we've dropped the lock and we will have suspended the `spa_async_thread`.  If we are unable to export then we re-enable the `spa_async_thread` but I don't see we do anything fix up the zvols. This isn't necessarily your bug but your change opens this window up.

I would also recommend running `zloop` for an extended period of time (if you haven't already) since it's good at catching these types of issues.